Lib/test/test_tabnanny.py

```diff
@@ -0,0 +1,353 @@
+                SOURCE_CODES["wrong_indented"], dir=tmp_dir
+            )
+            c_file = create_python_file(SOURCE_CODES["error_free"], dir=tmp_dir)
+            msg = ('unindent does not match any outer indentation level'
+                        ' (<tokenize>, line 3)\n')
+            msg = "%r: Indentation Error: %s" % (e_file, msg)
+            self._verify_tabnanny_check(tmp_dir, err=msg)
+
+
+class TestProcessTokens(TestCase):
+    """Testing `tabnanny.process_tokens()`."""
+
+    @patch('tabnanny.NannyNag')
+    def test_with_correct_code(self, MockNannyNag):
+        """A python source code without any whitespace related problems."""
+
+        file_path = create_python_file(SOURCE_CODES["error_free"])
+        with open(file_path) as f:
+            tabnanny.process_tokens(tokenize.generate_tokens(f.readline))
+        self.assertFalse(MockNannyNag.called)
+
+    def test_with_errored_codes_samples(self):
+        """A python source code with whitespace related sampled problems."""
+
+        # "tab_space_errored_1": exicutes block under type == tokenize.INDENT
+        #                        at `tabnanny.process_tokens()`.
+        # "tab space_errored_2": exicutes block under
+        #                        `check_equal and type not in JUNK` condition at
+        #                        `tabnanny.process_tokens()`.
+
+        for key in ["tab_space_errored_1", "tab_space_errored_2"]:
+            file_path = create_python_file(SOURCE_CODES[key])
+            exception = tabnanny.NannyNag
+            with open(file_path) as f:
+                tokens = tokenize.generate_tokens(f.readline)
+                self.assertRaises(exception, tabnanny.process_tokens, tokens)
+
+class TestCommandLine(TestCase):
+    """Tests command line interface of `tabnanny`."""
+
+    def _validate_cmd(self, *args, return_code=0, stdout=b"", stderr=b""):
+        """Common function to assert the behaviour of command line interface.
+
+        Arguments:
+            * args        : Add arguments which will be directly feeded as the
+                            arguments to tabnanny command line.
+            * return_code : Expected return code from tabnanny command line
+                            interface.
+            * stdout      : Expected standard output text after executing
+                            tabnanny command line interface.
+            * stderr      : Expected standard output error text after executing
+                            tabnanny command line interface.
+        """
+        rc, out, err = script_helper.assert_python_ok('-m', 'tabnanny', *args)
+        self.assertEqual(out, stdout)
+        self.assertEqual(rc, return_code)
+        self.assertEqual(err, stderr)
+
+    def test_with_errored_file(self):
+        """Should displays error when errored python file is given."""
+        file_path = create_python_file(SOURCE_CODES["wrong_indented"])
+        stderr = b"%r: Indentation Error: " % (file_path,)
```

Comment:
In Python 3, b"%r" is equivalent to calling the “ascii” function. I expect the test will fail if the file path has non-ASCII characters. It may be better to only test that an error is being reported, but tolerate variations in the message.